### ISSUE TITLE
parallel iqtree runs, 5% complete after 7 days

### ISSUE THREAD

**Issue by cmunro** [davidemms/OrthoFinder#349]
Firstly, thank you so much for building OrthoFinder, I am especially interested in using it after the recent update.

I am running OrthoFinder (v 2.3.8) on a 40 species dataset (mix of transcriptome/genome data), and hoping to generate gene trees via multiple sequence alignment (mafft) and ML inference (IQtree). For my downstream analyses I need good branch length estimates, and the default distance based tree estimates are not useful to me. However, I am finding that, as promised, iqtree is taking a long time to run -- so long, however, that I am not sure how practical it is to proceed. I was wondering if you might be able to provide some advice about optimising my runs. 

I am using 28 cores on a single node, and the following OrthoFinder command:
```
orthofinder -f [path]/sequences -t 28 -s [path]/speciestree.tree -M msa -T iqtree
```
I am running iqtree with a fixed model and no bootstraps (I tried running the default suggestion with ModelFinder, but this takes even longer). From the config json:
```
 "iqtree":{
    "program_type": "tree",
    "cmd_line": "iqtree -s INPUT -m LG+F+R4 -pre PATH/IDENTIFIER > /dev/null",
    "ouput_filename": "PATH/IDENTIFIER.treefile"
    }, 
```
I am currently 7 days in to the latest run, and in the `WorkingDirectory` I have 591 trees in `Trees_ids` and 595 in `Alignment_ids`. There is a total of 11042 orthogroups, so I still have a ways to go. I saw in another issue that I should expect the first 1000 or so to be slow because OrthoFinder starts with the largest orthogroups, but at this speed I would expect to be done with the first 1000 by the end of 2 weeks. Even if it does get faster after the first 1000, it could be months. 

As I understand it, the parallelisation is not at the iqtree level but at the OrthoFinder level - so msa an phylogenetic inference runs are occurring in parallel but each only using a single core. I looked at the iqtree log files, and some of the first trees are still running, while others have completed their runs after 72+ hours. 

Is there any way to optimise this? Thank you so much for your help.

All the best,
Catriona

**Comment by cmunro** [comment]
This is perhaps an unrelated error, but while going through log files I saw the following error:
```Reading alignment file ~/OrthoFinder/sequences/OrthoFinder/Results_Feb16/WorkingDirectory/Alignments_ids/OG0000000.fa ... ERROR: Cannot read file ~/OrthoFinder/sequences/OrthoFinder/Results_Feb16/WorkingDirectory/Alignments_ids/OG0000000.fa
``
Is this a mafft issue? OG0000000.fa is completely empty. I noticed several core dumps in my working directory.

**Comment by davidemms** [comment]
Hi Catriona

I think you've specified most of the major considerations. I've very rarely run IQTREE for all orthogroups, and when I have it's only been on the smallest datasets because the run time is very high. I would say:

* If you've done 591 trees then that's relatively promising. The largest orthogroups take by far the longest time and so it will be accelerate considerably at this point. If you're wondering whether to continue then it is worth monitoring the acceleration in order to estimate when it might finish. I find `ls -lct` on the Tree_ids/ directory is useful as you can see how long the first 100 took, the next 100 etc. You may be nearer 25% of the way through, as a bit of a wild guess, rather than 5%. For this reason it's probably better continuing with this run than making optimisations and starting again.
* I'm less sure on this but I think +F may slow it down more than necessary and could be dropped. At a push maybe even the +G4. I would take something like OG 1000 or OG 500 and see how long it takes with and without these IQTREE model options.
* RAM is also a problem with multiple IQTREE runs as it's usage is quite high. I assume you must have a lot of RAM if you're able to run 28 in parallel? It may be worth checking you're not going into swap as this will obviously slow things down a lot. Also, check that none of your iqtree runs have crashed without completing. (You could, for example, grep the iqtree "*.log" files for the a line that indicates they have finished e.g. "Analysis results written to"). At most 28 of your trees should still be in progress and therefore currently missing this line (e.g. `grep "Analysis results written to" *log | wc -l`)
* The parallelisation over trees (one core each) will give perfect parallelisation whereas this would be an upper limit for what iqtree could obtain, so this is the best approach to take.
* The error you posted suggests that the alignment failed for OG 0. Often you won't be particularly interested in this because it will be virtually impossible to work out the relationships between the genes in this orthogroup. It may be better to not worry too much about this one, and if you need orthologs or the tree then just use the results from a default orthofinder run. Alternatively, you can try to get it to run manually and use that.
* It would also be possible to trim alignments before tree inference, which would make things faster. I haven't made this an option because I can't find any automated trimming tools which can guarantee that you'll get a good (or even non-zero length) alignment out in order for the tree inference to be successful. [Noisy](Noisy: Identification of problematic columns in multiple sequence alignments) was the best I found, but there were problems. Trimal can trim all (!) of the sequence away.

All the best
David

**Comment by cmunro** [comment]
Hi David, 

Thank you so much for your thorough response.  

- It seems that your wild guess on percentage complete was much better than mine. I checked last night and I was already at 2833 trees in the `Trees_ids` folder. I checked again this morning and it's at 10471 trees. I hit that 1000 tree mark and the runs sped up significantly.

- You're probably right that if I dropped +G and +F it would probably be faster.  Exploring parameter space for just one or two OG trees is probably a good way to go, thank you for the suggestion. 

- I ran `grep "Analysis results written to" *log | wc -l` (result: 10466) and also `grep "SIGNAL BUS ERROR" *log | wc -l` (result: 4). I supplied 60G of memory.  It seems like the vast majority of trees are running just fine, but there are a few that are running into issues.

- It looks like OG 0 isn't the only one, there's also OG 14 and 8 that are empty alignments (grep "Cannot read file" *log | wc -l`. These are both likely to be large orthogroups, so perhaps this is also to be expected. 

- Alignment trimming would speed things up, but I don't have any good suggestions for programs that don't generate non-zero length alignments. I was previously inputting Gblocks trimmed alignments from [Agalma](https://bitbucket.org/caseywdunn/agalma/src/master/) into [PHYLDOG](https://github.com/Boussau/PHYLDOG). I checked with Agalma developer Felipe Zapata and he said they did sometimes run into zero-length alignments that they skipped over when moving into gene tree inference stages. I would be curious to check how many trees would be lost with automated trimming. 

There are several large OG alignments that have been running since I started (now 9 days ago) but are still at very early stages. I guess at some point I might have to kill these jobs and try to re-run them separately, likewise with to re-run those that have errors. What is the appropriate approach for downstream analysis? 
I expect I would:
- Run iqtree on the relevant alignments in `Alignments_ids`
- Move the tree file to `Trees_ids` and append `tree_id.txt`
The `-ft` command will be looking for trees moved to the Results directory, I suppose. Can I supply trees from the WorkingDirectory? Or do I need to change the gene names and move the tree to the Results directory. If so, where can I find a list of the original names and corresponding temporary names?

Thanks again for your very helpful response. 

Catriona

**Comment by davidemms** [comment]
Hi Catriona

You're welcome! OrthoFinder just requires the versions of the tree files in the WorkingDirectory so your proposed workflow should work, let me know if you have any problems. To ensure that it works smoothly if you do go down this root: the best way to get there would be to kill the individual iqtree or mafft processes that are taking too long rather than the orthofinder process. OrthoFinder should then be able to continue running to completion that way, and just ignore the ones that failed. This will just ensure that any other files it needs are in the correct places for restarting it with '-ft'. Also when restarting, remember to supply your species tree again.

All the best
David

**Comment by cmunro** [comment]
Hi David,

The run completed! It took 11 days in the end. It seems that the ongoing iqtree runs were killed without me needing to do anything. In the end, I got 10095 trees (out of 11041 orthogroups).  In terms of the 946 orthogroups remaining, 580 orthogroups failed because they were too small (2 species), 4 failed at the alignment stage (possibly too large?), 4 crashed before completing. I'm not sure about the remaining 358 trees, but I suspect these all timed out before completion.  I will try to re-run the 4 crashed trees and the 358, in this case supplying more cores per process to see it through. 

I'm glad to hear that I can use tree files in the WorkingDirectory, this is very helpful. I'll definitely let you know if I run into any issues. 

All the best,
Catriona

**Comment by awesomedeer** [comment]
Hi David!
How to check whether I am going to SWAP? I check my iqtree commands they are like less than 10% memory usage. Is it normal or it means I am going to SWAP?
Best regards
Song
